skimage/morphology/gray.py

```diff
@@ -0,0 +1,489 @@
+
+    Returns
+    -------
+    closing : array, same shape and type as `image`
+        The result of the morphological closing.
+
+    Examples
+    --------
+    >>> # Close a gap between two bright lines
+    >>> import numpy as np
+    >>> from skimage.morphology import square
+    >>> broken_line = np.array([[0, 0, 0, 0, 0],
+    ...                         [0, 0, 0, 0, 0],
+    ...                         [1, 1, 0, 1, 1],
+    ...                         [0, 0, 0, 0, 0],
+    ...                         [0, 0, 0, 0, 0]], dtype=np.uint8)
+    >>> closing(broken_line, square(3))
+    array([[0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0],
+           [1, 1, 1, 1, 1],
+           [0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0]], dtype=uint8)
+
+    """
+    dilated = dilation(image, selem)
+    # note: shift_x, shift_y do nothing if selem side length is odd
+    out = erosion(dilated, selem, out=out, shift_x=True, shift_y=True)
+    return out
+
+
+@default_selem
+def white_tophat(image, selem=None, out=None):
+    """Return white top hat of an image.
+
+    The white top hat of an image is defined as the image minus its
+    morphological opening. This operation returns the bright spots of the image
+    that are smaller than the structuring element.
+
+    Parameters
+    ----------
+    image : ndarray
+        Image array.
+    selem : ndarray, optional
+        The neighborhood expressed as an array of 1's and 0's.
+        If None, use cross-shaped structuring element (connectivity=1).
+    out : ndarray, optional
+        The array to store the result of the morphology. If None
+        is passed, a new array will be allocated.
+
+    Returns
+    -------
+    out : array, same shape and type as `image`
+        The result of the morphological white top hat.
+
+    See Also
+    --------
+    black_tophat
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Top-hat_transform
+
+    Examples
+    --------
+    >>> # Subtract gray background from bright peak
+    >>> import numpy as np
+    >>> from skimage.morphology import square
+    >>> bright_on_gray = np.array([[2, 3, 3, 3, 2],
+    ...                            [3, 4, 5, 4, 3],
+    ...                            [3, 5, 9, 5, 3],
+    ...                            [3, 4, 5, 4, 3],
+    ...                            [2, 3, 3, 3, 2]], dtype=np.uint8)
+    >>> white_tophat(bright_on_gray, square(3))
+    array([[0, 0, 0, 0, 0],
+           [0, 0, 1, 0, 0],
+           [0, 1, 5, 1, 0],
+           [0, 0, 1, 0, 0],
+           [0, 0, 0, 0, 0]], dtype=uint8)
+
+    """
+    selem = np.array(selem)
+    if out is image:
+        opened = opening(image, selem)
+        if np.issubdtype(opened.dtype, bool):
+            np.logical_xor(out, opened, out=out)
+        else:
+            out -= opened
+        return out
+    elif out is None:
+        out = np.empty_like(image)
+    # work-around for NumPy deprecation warning for arithmetic
```

Comment:
It would be good to have a reference for the specific deprecation cycle.  Indeed, maybe this is obsolete by now?

Reply:
The deprecation cycle is complete and this is now a TypeError (not positive the minimum NumPy where this occured). I will update the message. 

example
```Python
np.ones(8, dtype=bool) - np.ones(8, dtype=bool)
```
raises
```
TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
```